docs/design/features/byreflike-generics.md

```diff
@@ -0,0 +1,76 @@
+- `constrained.callvirt` &ndash; This IL sequence must resolve to a method implemented on `object`, or a default interface method.
+
+If any of the above instructions are attempted to be used with a ByRefLike type, the runtime will throw an `InvalidProgramException`.
+
+The following instructions are already set up to support this feature since their behavior will fail as currently defined due to the inability to box a ByRefLike type.
+
+- `throw` &ndash; Requires an object reference to be on stack, which can never be a ByRefLike type.
+- `unbox` / `unbox.any` &ndash; Requires an object reference to be on stack, which can never be a ByRefLike type.
+- `isinst` &ndash; Will always place `null` on stack.
+- `castclass` &ndash; Will always throw `System.InvalidCastException`.
+
+## Proposal
+
+Support for the following will be indicated by the `RuntimeFeature.GenericsAcceptByRefLike` property.
+
+A new `GenericParameterAttributes` value will be defined which also represents metadata defined in the `CorGenericParamAttr` enumeration. Space is provided between the existing constraints group to permit constraint growth.
+
+```diff
+namespace System.Reflection
+{
+    [Flags]
+    public enum GenericParameterAttributes
+    {
++        AcceptByRefLike = 0x0100
+    }
+}
+```
+
+```diff
+typedef enum CorGenericParamAttr
+{
++   gpAcceptByRefLike = 0x0100 // type argument can be ByRefLike
+} CorGenericParamAttr;
+```
+
+The expansion of metadata will impact at least the following:
+
+- ILDasm/ILAsm &ndash; https://github.com/dotnet/runtime
+- Cecil &ndash; https://github.com/jbevain/cecil
+- IL Trimmer &ndash; https://github.com/dotnet/linker
+- F# &ndash; https://github.com/fsharp/fsharp
+- C++/CLI &ndash; The MSVC team
+
+An API that is a JIT-time intrinsic will be needed to determine if a parameter is ByRefLike. This API would represent a check to occur at JIT time code-gen to avoid taking paths that would be invalid for some values of `T`. The existing `Type.IsByRefLike` property will be made an intrinsic (e.g., `typeof(T).IsByRefLike`).
+
+For dispatch to object implemented methods and to default interface methods, the behavior shall be that an `InvalidProgramException` should be thrown. The JIT would insert the following IL at code-gen time.
+
+```
+newobj instance void System.InvalidProgramException::.ctor()
+throw
+```
+
+When boxing due to a constrained call that cannot be made, instead of allocating a normal boxed object, an object of `InvalidBoxedObject` type will be created. It will have implementations of the various overridable object methods which throw `InvalidProgramException`, and interface dispatch shall have a special case for attempting to invoke an interface method on such an object, that will also throw an `InvalidProgramException`.
+
+The `Reflection.Emit` API will need to be updated to respect the behavior of this flag. How it will handle support is an open question.
+
+## Open questions
```

Comment:
There are other proposals out there to expand the set of allowed generic arguments. Most notably to include pointers. Do you anticipate this design will cause future compat issues with those proposals?

Reply:
I can't see that causing an issue. We have a fair number of bits (10) remaining on `GenericParameterAttributes` after this one so we are still good to support both. I suppose we could look at devising a more extensive mechanism if there are future concerns.